# Exception

- [Exception](#exception)
  - [异常的作用](#异常的作用)
  - [Try catch](#try-catch)
  - [异常分类](#异常分类)
  - [如何自定义一个异常类？](#如何自定义一个异常类)

## 异常的作用

异常机制可以帮助我们定位错误,给出错误出现的原因,帮助我们解决问题,异常机制增强了程序的健壮性、可维护性

## Try catch

- Try catch  try检测出异常后执行暂停 抛出给catch(如无匹配则抛给jvm)
- catch内部代码执行后  执行catch块下面的代码(而不是报告给 jvm终止程序)
- finally一定会被执行 即使有return 普通代码不执行 finally也会执行
- finally是在return表达式运行后执行的，此时将try或catch中的return值或表达式执行一次但是结果暂存，待finally代码块执行结束后再将之前暂存的结果返回

```java
// 无错误执行结果为8,错误为10
static int a(){
    int i  = 0;
    try {
        i = i/0;
        return i+= 2;
    } catch (Exception e) {
        e.printStackTrace();
        return i+= 4;
    } finally {
        return i+= 6;}}
```

## 异常分类

Throwable

1. Exception
    1. IOException(checked)
        1. EOFException
        2. FileNotFoundException
        3. SocketException
        4. FileSystemException
        5. HttpRetryException
        6. ZipException
    2. RuntimeExcetion
        1. ArithmeticException
        2. ClassCastException
        3. IllegalArgumentException
        4. IllegalStateException
        5. IndexOutOfBoundsException
        6. NoSuchElementException
        7. NullPointerException
        8. ConcurrentModificationException
        9. MalformedURLException
        10. UnknownHostException
        11. ClassNotFoundException
        12. CloneNotSupportedException
    3. AWTException
2. Error
    1. VirtulMachineError
        1. StackOverFlowError
        2. OutOfMemoryError
        3. InternalError
    2. AWTError
    3. LinkageError
        1. BootstrapMethodError
        2. NoClassDefFoundError
        3. VerifyError

Error报错，会交给JVM，throw出错误并终止程序

## 如何自定义一个异常类？

```java
/*
    * 如何自定义异常类？
    * 1. 继承于现的异常结构：RuntimeException 、Exception
    * 2. 提供全局常量：serialVersionUID
    * 3. 提供重载的构造器
    * 
    */
public class MyException extends Exception{
    
    static final long serialVersionUID = -7034897193246939L;
    
    public MyException(){}

    public MyException(String msg){
        super(msg);}}
```
